### PR TITLE
[DO NOT MERGE] Add IOperation APIs for string interpolation

### LINKED ISSUE #18300: IOperation doesn't support string interpolations
**Example Code** 
using System;
class P
{
    private void M()
    {
        var s = "";
        if (int.TryParse(s, out var i))
            Console.WriteLine($"i ={ i}, s ={ s}");
        else
          Console.WriteLine($"i ={ i}, s ={ s}");
    }
}

Try generating Ioperation tree for Ifstatement
**Generated Tree**

IOperation tree for "if (int.TryParse(s, out var i))
            Console.WriteLine($"i ={ i}, s ={ s}");
        else
          Console.WriteLine($"i ={ i}, s ={ s}");"

IIfStatement (OperationKind.IfStatement)
  Condition: IInvocationExpression (static System.Boolean System.Int32.TryParse(System.String s, out System.Int32 result)) (OperationKind.InvocationExpression, Type: System.Boolean)
      IArgument (Matching Parameter: s) (OperationKind.Argument)
        ILocalReferenceExpression: s (OperationKind.LocalReferenceExpression, Type: System.String)
      IArgument (Matching Parameter: result) (OperationKind.Argument)
        ILocalReferenceExpression: i (OperationKind.LocalReferenceExpression, Type: System.Int32)
  IExpressionStatement (OperationKind.ExpressionStatement)
    IInvocationExpression (static void System.Console.WriteLine(System.String value)) (OperationKind.InvocationExpression, Type: System.Void)
      IArgument (Matching Parameter: value) (OperationKind.Argument)
        IOperation:  (OperationKind.None)
  IExpressionStatement (OperationKind.ExpressionStatement)
    IInvocationExpression (static void System.Console.WriteLine(System.String value)) (OperationKind.InvocationExpression, Type: System.Void)
      IArgument (Matching Parameter: value) (OperationKind.Argument)
        IOperation:  (OperationKind.None)

### PR DESCRIPTION
**Pull request by mavasani**
This is a proposal PR for https://github.com/dotnet/roslyn/issues/18300#issuecomment-294631382. We will discuss the proposal at the next IOperation Design meeting. Subsequently, I will add unit tests for string interpolation operation trees.

Fixes #18300 

**Customer scenario**

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

**Bugs this fixes:**

(either VSO or GitHub links)

**Workarounds, if any**

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

**Risk**

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

**Performance impact**

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

**Is this a regression from a previous update?**

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

**How was the bug found?**

(E.g. customer reported it vs. ad hoc testing)

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml [hunk 1]
```diff
@@ -1897,13 +1897,13 @@
...
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
-    <Field Name="Contents" Type="ImmutableArray(Of BoundNode)"/>
+    <Field Name="Contents" Type="ImmutableArray(Of BoundExpression)"/>
     <Field Name="Binder" Type="Binder" Null="disallow" />
   </Node>
   
   <!-- There is no BoundInterpolatedStringText because we use BoundLiteral of type string for this purpose. -->
   
-  <Node Name="BoundInterpolation" Base="BoundNode">
+  <Node Name="BoundInterpolation" Base="BoundExpression" HasValidate="true">
```
Comment (mavasani) [inline src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml]: See https://github.com/dotnet/roslyn/issues/18300#issuecomment-294983713

**Comment by mavasani** [comment]
Tagging @dotnet/analyzer-ioperation @AlekseyTs @cston for an early API review.

**Comment by AlekseyTs** [comment]
CC @gafter, @AnthonyDGreen Since you, guys, implemented the feature for C# and VB, could you please take a look at the proposed design?

File: src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml]: Please revert this change.

File: src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml [hunk 2]
```diff
@@ -1897,13 +1897,13 @@
   <Node Name="BoundInterpolatedStringExpression" Base="BoundExpression" HasValidate="true">
     <!-- Type is required for this node type; may not be null -->
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
-    <Field Name="Contents" Type="ImmutableArray(Of BoundNode)"/>
+    <Field Name="Contents" Type="ImmutableArray(Of BoundExpression)"/>
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml]: Please revert this change.

File: src/Compilers/VisualBasic/Portable/Binding/Binder_InterpolatedString.vb [hunk 3]
```diff
@@ -65,7 +65,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                                   CreateStringLiteral(syntax.FormatClause, syntax.FormatClause.FormatStringToken.ValueText, compilerGenerated:=False, diagnostics:=diagnostics),
                                   Nothing)
 
-            Return New BoundInterpolation(syntax, expression, alignmentOpt, formatStringOpt)
+            Return New BoundInterpolation(syntax, expression, alignmentOpt, formatStringOpt, type:=Nothing)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder_InterpolatedString.vb]: >Return New BoundInterpolation(syntax, expression, alignmentOpt, formatStringOpt, type:=Nothing) [](start = 12, length = 95)

Please revert this change.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml (hunk 1 above)
Comment (mavasani) [inline src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml]: Do we still want consistency between C# and VB here? If so, we should change BoundInterpolation for C# to be a BoundNode.

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml]: > Do we still want consistency between C# and VB here?

I am fine with leaving things as they are today, but wouldn't object to changing C# implementation to be more like VB. It is not clear to me whether and how exactly this affects IOperation.

Comment (AnthonyDGreen) [inline src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml]: They aren't expressions. We could add some base types and make the text content an explicit type but I think it would be dangerous to imply these are always expressions. It's more of an implementation shortcut that I reused BoundLiteral there.

**Comment by mavasani** [comment]
This was implemented in another PR, closing this one.
